model/project_ref.go

```diff
@@ -530,6 +518,171 @@ func recursivelySetUndefinedFields(structToSet, structToDefaultFrom reflect.Valu
+		}
+	}
+}
+
+func (p *ProjectRef) createNewRepoRef(u *user.DBUser) (repoRef *RepoRef, err error) {
+	repoRef = &RepoRef{ProjectRef{
+		Id:      mgobson.NewObjectId().Hex(),
+		Owner:   p.Owner,
+		Repo:    p.Repo,
+		Enabled: utility.TruePtr(),
+		Admins:  []string{},
+	}}
+
+	allEnabledProjects, err := FindMergedEnabledProjectRefsByOwnerAndRepo(p.Owner, p.Repo)
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding all enabled projects")
+	}
+	// for every setting in the project ref, if all enabled projects have the same setting, then use that
+	defer func() {
+		err = recovery.HandlePanicWithError(recover(), err, "project and repo structures do not match")
+	}()
+	setRepoFieldsFromProjects(repoRef, allEnabledProjects)
+	repoRef.Admins = append(repoRef.Admins, u.Username())
+
+	// creates scope and give user admin access to repo
+	if err = repoRef.Add(u); err != nil {
+		return nil, errors.Wrapf(err, "problem adding new repo repo ref for '%s/%s'", p.Owner, p.Repo)
+	}
+	enabledProjectIds := []string{}
+	for _, p := range allEnabledProjects {
+		enabledProjectIds = append(enabledProjectIds, p.Id)
+	}
+	commonProjectVars, err := getCommonProjectVariables(enabledProjectIds)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting common project variables")
+	}
+	commonProjectVars.Id = repoRef.Id
+	if err = commonProjectVars.Insert(); err != nil {
+		return nil, errors.Wrap(err, "error inserting project variables for repo")
+	}
+
+	commonAliases, err := getCommonAliases(enabledProjectIds)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting common project aliases")
+	}
+	for _, a := range commonAliases {
+		a.ProjectID = repoRef.Id
+		if err = a.Upsert(); err != nil {
+			return nil, errors.Wrapf(err, "error upserting alias for repo")
+		}
+	}
+
+	return repoRef, nil
+}
+
+func getCommonAliases(projectIds []string) (ProjectAliases, error) {
+	commonAliases := map[string]ProjectAlias{} // use map for easy removal of non-common aliases
+	for i, id := range projectIds {
+		aliases, err := FindAliasesForProject(id)
+		if err != nil {
+			return nil, errors.Wrap(err, "error finding aliases for project")
+		}
+		if i == 0 {
+			for _, a := range aliases {
+				commonAliases[a.ID.Hex()] = a
+			}
+			continue
+		}
+		for _, commonAlias := range commonAliases {
+			// look to see if this alias exists in the each project and if not remove it
+			if !aliasSliceContains(aliases, commonAlias) {
+				delete(commonAliases, commonAlias.ID.Hex())
+			}
+		}
+		if len(commonAliases) == 0 {
+			return nil, nil
+		}
+	}
+
+	res := ProjectAliases{}
+	for _, a := range commonAliases {
+		res = append(res, a)
+	}
+	return res, nil
+}
+
+func aliasSliceContains(slice []ProjectAlias, item ProjectAlias) bool {
+	for _, each := range slice {
+		if each.RemotePath == item.RemotePath && each.Alias == item.Alias && each.GitTag == item.GitTag &&
+			each.Variant == item.Variant && reflect.DeepEqual(each.VariantTags, item.VariantTags) &&
```

Comment:
I would use StringSliceIntersection for these since it's less likely to panic and also probably is faster